File: src/main/kotlin/com/example/features/InfixAndOperatorDemo.kt
package com.example.features

// ─── 2D vector class with full operator overloads ────────────────────────────

/**
 * Simple 2-D integer vector that demonstrates every operator overload
 * supported by the transpiler.  Each operator emits a Lua metamethod so
 * that the standard Lua arithmetic operators (+, -, *) dispatch correctly.
 *
 * All property accesses use explicit `this.` so the transpiler renders
 * them as `self.x`, `self.y`, etc. in Lua.
 */
class Vec2(val x: Int, val y: Int) {
    /** a + b → Lua __add */
    operator fun plus(other: Vec2): Vec2 = Vec2(this.x + other.x, this.y + other.y)

    /** a - b → Lua __sub */
    operator fun minus(other: Vec2): Vec2 = Vec2(this.x - other.x, this.y - other.y)

    /** a * scalar → Lua __mul */
    operator fun times(scalar: Int): Vec2 = Vec2(this.x * scalar, this.y * scalar)

    /** -a → Lua __unm */
    operator fun unaryMinus(): Vec2 = Vec2(-this.x, -this.y)

    /** a == b → Lua __eq */
    override operator fun equals(other: Any?): Boolean =
        other is Vec2 && this.x == other.x && this.y == other.y

    override fun hashCode(): Int = 31 * this.x + this.y

    /** magnitude-squared, used by compareTo */
    fun mag2(): Int = this.x * this.x + this.y * this.y

    /** a < b, a <= b → Lua __lt and __le */
    operator fun compareTo(other: Vec2): Int = this.mag2() - other.mag2()

    /** tostring(v) → Lua __tostring (emitted for override toString) */
    override fun toString(): String = "(${this.x}, ${this.y})"
}

/**
 * Callable function-object that multiplies its argument by a fixed factor.
 * Demonstrates `operator fun invoke` → Lua `__call` metamethod.
 */
class Multiplier(val factor: Int) {
    operator fun invoke(x: Int): Int = this.factor * x
}

class CoolCat(val name: String) {
    var coolPoints: Int = 0
    fun sayHello() = println("Hello, I'm $name!")
    operator fun plusAssign(amount: Int) { this.coolPoints += amount }
    operator fun minusAssign(amount: Int) { this.coolPoints -= amount }
}

/**
 * A simple accumulator demonstrating all five compound-assignment operator overloads.
 * Each overload mutates `value` in-place.
 */
class Accumulator(var value: Int) {
    operator fun plusAssign(n: Int)  { this.value += n }
    operator fun minusAssign(n: Int) { this.value -= n }
    operator fun timesAssign(n: Int) { this.value *= n }
    operator fun divAssign(n: Int)   { this.value /= n }
    operator fun remAssign(n: Int)   { this.value %= n }
}

// ─── Extension function ───────────────────────────────────────────────────────

fun CoolCat.printName() = println(this.name)

/** Extension property: a greeting string using the cat's name. */
val CoolCat.greeting: String
    get() = "Hello, I'm ${this.name}!"

/** Appends " !" to the receiver string.  Transpiled as `shout(self)`. */
fun String.shout(): String = this + " !"

// ─── Extension properties ─────────────────────────────────────────────────────

/**
 * Extension property: the number of words in a string (split on spaces).
 * Transpiled as a getter function `wordCount(self)` in Lua.
 * Accessed as `s.wordCount` → `wordCount(s)`.
 */
val String.wordCount: Int
    get() = this.split(" ").size

/**
 * Extension property: the last character of a string as a one-char string.
 * Transpiled as a getter function `lastChar(self)` in Lua.
 */
val String.lastChar: String
    get() = this.substring(this.length - 1)

// ─── Custom infix function ────────────────────────────────────────────────────

/** Returns true when [this] is strictly between [lo] and [hi]. */
infix fun Int.isBetween(range: Pair<Int, Int>): Boolean =
    this > range.first && this < range.second

// ─── Demo functions ───────────────────────────────────────────────────────────

/**
 * Exercises Vec2 operator overloads: +, -, *, unary -, compareTo, toString.
 * Returns a checksum so the call is not dead-code.
 */
fun demoVec2Operators(): Int {
    val a = Vec2(3, 4)          // mag2 = 25
    val b = Vec2(1, 2)          // mag2 = 5
    val c = a + b               // Vec2(4, 6)
    val d = a - b               // Vec2(2, 2)
    val e = b * 3               // Vec2(3, 6)
    val neg = -b                // Vec2(-1, -2)

    // Comparison via compareTo (mag2-based)
    val aGreater = a.compareTo(b) > 0   // 25 - 5 = 20 > 0 → true
    val bSmaller = b.compareTo(a) < 0   // 5 - 25 = -20 < 0 → true

    return c.x + d.x + e.x + neg.x +
        (if (aGreater) 1 else 0) + (if (bSmaller) 1 else 0)
    // (4) + (2) + (3) + (-1) + 1 + 1 = 10
}

/**
 * Exercises the `invoke` operator overload via [Multiplier].
 */
fun demoInvokeOperator(): Int {
    val triple = Multiplier(3)
    return triple.invoke(7)   // 21
}

/**
 * Exercises extension functions: [String.shout].
 */
fun demoExtensionFunction(): String {
    val word = "hello"
    return word.shout()   // "hello !"
}

/**
 * Exercises custom infix function [isBetween].
 */
fun demoCustomInfix(): Int {
    val x = 5
    val inRange = x isBetween Pair(0, 10)   // true: 5 > 0 and 5 < 10
    val notIn   = x isBetween Pair(5, 10)   // false: 5 is not > 5
    return (if (inRange) 1 else 0) + (if (notIn) 1 else 0)  // 1
}

/**
 * Exercises `with(x) { ... }` scope function: the receiver is bound to `self`
 * inside the block.
 */
fun demoWith(): Int {
    val v = Vec2(10, 20)
    return with(v) {
        this.x + this.y   // explicit this → self.x + self.y in Lua
    }
}

/**
 * Exercises `to` as a method call.
 */
fun demoToMethod(): List<Any> {
    val pair = "key" to 42   // {key, 42} in Lua
    return listOf(pair)
}

/**
 * Exercises collection binary `+` and `-` operators.
 * When the left-hand side is a literal constructor the transpiler routes to
 * `ktox_plus` / `ktox_minus`.  For map + pair, `ktox_mapPlus` is emitted.
 */
fun demoCollectionOperators(): Int {
    // list + list → ktox_plus (both operands are list constructors)
    val combined = listOf(1, 2, 3) + listOf(4, 5)   // {1,2,3,4,5}
    // list - element → ktox_minus
    val removed  = listOf(1, 2, 3, 4, 5) - 3        // {1,2,4,5}
    // map + pair → ktox_mapPlus (left is map constructor)
    val m = mapOf("x" to 1) + ("y" to 2)            // {["x"]=1,["y"]=2}
    // Use containsKey to verify map entries (Lua # does not count string keys)
    val hasBoth = m.containsKey("x") && m.containsKey("y")
    return combined.size + removed.size + (if (hasBoth) 1 else 0)
    // 5 + 4 + 1 = 10
}

/**
 * Exercises `matches` (both dot-call and infix forms with Regex).
 * The transpiler emits `ktox_matches(s, pattern)` for both forms.
 */
fun demoMatches(): Int {
    val s = "hello123"
    val pattern = Regex("[a-z]+[0-9]+")
    val dotForm   = s.matches(pattern)                    // true
    val infixForm = s matches Regex("[a-z]+[0-9]+")       // true
    val noMatch   = s matches Regex("[0-9]+")              // false
    return (if (dotForm) 1 else 0) + (if (infixForm) 1 else 0) + (if (noMatch) 0 else 1)
    // 1 + 1 + 1 = 3
}

/**
 * Exercises extension properties [String.wordCount] and [String.lastChar].
 * The transpiler emits getter function calls: `wordCount(s)` and `lastChar(s)`.
 * Returns 8 = wordCount("hello world foo") + lastChar_is_o + wordCount("one") + lastChar_is_e
 */
fun demoExtensionProperty(): Int {
    val phrase = "hello world foo"   // 3 words
    val wc = phrase.wordCount        // wordCount(phrase) → 3
    val lc = phrase.lastChar         // lastChar(phrase) → "o"

    val single = "one"
    val wc2 = single.wordCount       // wordCount(single) → 1
    val lc2 = single.lastChar        // lastChar(single) → "e"

    return wc + (if (lc == "o") 1 else 0) + wc2 + (if (lc2 == "e") 1 else 0)
    // 3 + 1 + 1 + 1 = 6... wait: 3+1+1+1 = 6
}

fun demoExtensionFunctionOnUserDefined() {
    val awesomeCat = CoolCat("teacup")
    awesomeCat.printName()
}

/**
 * Exercises the class extension property [CoolCat.greeting].
 * Returns the greeting string so the test can verify its value.
 */
fun demoClassExtensionProperty(): String {
    val cat = CoolCat("Whiskers")
    return cat.greeting
}

/**
 * Exercises the [CoolCat.plusAssign] and [CoolCat.minusAssign] operator overloads.
 * Returns the coolPoints after adding 10, 5, then subtracting 3 (expected: 12).
 */
fun demoCoolCatPlusAssign(): Int {
    val cat = CoolCat("Mittens")
    cat += 10
    cat += 5
    cat -= 3
    return cat.coolPoints
}

/**
 * Exercises all five compound-assignment operator overloads on [Accumulator].
 * Start at 10 → +5=15 → -3=12 → *2=24 → /4=6 → %5=1 (expected: 1).
 */
fun demoAllCompoundAssignOverloads(): Int {
    val acc = Accumulator(10)
    acc += 5   // 15
    acc -= 3   // 12
    acc *= 2   // 24
    acc /= 4   // 6
    acc %= 5   // 1
    return acc.value
}

File: outputDir/vscripts/features/InfixAndOperatorDemo.lua
-- package: com.example.features

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "features/InfixAndOperatorDemo.kt", {["1-90"]=1,["91-94"]=53,["95-108"]=54,["109-112"]=62,["113-116"]=63,["117-120"]=64,["121-124"]=65,["125-153"]=66,["154"]=110,["155"]=111,["156"]=112,["157"]=113,["158"]=114,["159"]=115,["160"]=118,["161"]=119,["162-165"]=121,["166"]=130,["167-170"]=131,["171"]=138,["172-175"]=139,["176"]=146,["177"]=147,["178"]=148,["179-182"]=149,["183"]=157,["184-189"]=158,["190"]=167,["191-194"]=168,["195"]=178,["196"]=180,["197"]=182,["198"]=184,["199-202"]=185,["203"]=194,["204"]=195,["205"]=196,["206"]=197,["207"]=198,["208-211"]=199,["212"]=209,["213"]=210,["214"]=211,["215"]=213,["216"]=214,["217"]=215,["218-221"]=217,["222"]=222,["223-226"]=223,["227"]=231,["228-231"]=232,["232"]=240,["233"]=241,["234"]=242,["235"]=243,["236-239"]=244,["240"]=252,["241"]=253,["242"]=254,["243"]=255,["244"]=256,["245"]=257,["246-248"]=258})

Vec2 = {}
Vec2.__index = Vec2
Vec2.__add = function(a, b) return a:plus(b) end
Vec2.__sub = function(a, b) return a:minus(b) end
Vec2.__mul = function(a, b) return a:times(b) end
Vec2.__unm = function(a) return a:unaryMinus() end
Vec2.__eq = function(a, b) return a:equals(b) end
Vec2.__lt = function(a, b) return a:compareTo(b) < 0 end
Vec2.__le = function(a, b) return a:compareTo(b) <= 0 end
Vec2.__tostring = function(a) return a:toString() end

function Vec2:new(x, y)
    local self = setmetatable({}, Vec2)
    self.x = x
    self.y = y
    return self
end

function Vec2:plus(other)
    return Vec2:new(self.x + other.x, self.y + other.y)
end

function Vec2:minus(other)
    return Vec2:new(self.x - other.x, self.y - other.y)
end

function Vec2:times(scalar)
    return Vec2:new(self.x * scalar, self.y * scalar)
end

function Vec2:unaryMinus()
    return Vec2:new(-self.x, -self.y)
end

function Vec2:equals(other)
    return ktox_isinstance(other, Vec2) and self.x == other.x and self.y == other.y
end

function Vec2:hashCode()
    return 31 * self.x + self.y
end

function Vec2:mag2()
    return self.x * self.x + self.y * self.y
end

function Vec2:compareTo(other)
    return self:mag2() - other:mag2()
end

function Vec2:toString()
    return "(" .. tostring(self.x) .. ", " .. tostring(self.y) .. ")"
end


Multiplier = {}
Multiplier.__index = Multiplier
Multiplier.__call = function(a, ...) return a:invoke(...) end

function Multiplier:new(factor)
    local self = setmetatable({}, Multiplier)
    self.factor = factor
    return self
end

function Multiplier:invoke(x)
    return self.factor * x
end


CoolCat = {}
CoolCat.__index = CoolCat

function CoolCat:new(name)
    local self = setmetatable({}, CoolCat)
    self.name = name
    self.coolPoints = 0
    return self
end

function CoolCat:sayHello()
    return println("Hello, I\'m " .. tostring(name) .. "!")
end

function CoolCat:plusAssign(amount)
    self.coolPoints = ktox_plusAssign(self.coolPoints, amount)
end

function CoolCat:minusAssign(amount)
    self.coolPoints = ktox_minusAssign(self.coolPoints, amount)
end


Accumulator = {}
Accumulator.__index = Accumulator

function Accumulator:new(value)
    local self = setmetatable({}, Accumulator)
    self.value = value
    return self
end

function Accumulator:plusAssign(n)
    self.value = ktox_plusAssign(self.value, n)
end

function Accumulator:minusAssign(n)
    self.value = ktox_minusAssign(self.value, n)
end

function Accumulator:timesAssign(n)
    self.value = self.value * n
end

function Accumulator:divAssign(n)
    self.value = self.value / n
end

function Accumulator:remAssign(n)
    self.value = self.value % n
end


function CoolCat:printName()
    return println(self.name)
end

function CoolCat:greeting()
    return "Hello, I\'m " .. tostring(self.name) .. "!"
end

function string_shout(self)
    return self .. " !"
end

function string_wordCount(self)
    return #ktox_split(self, " ")
end

function string_lastChar(self)
    return ktox_substring(self, #self - 1)
end

function int_isBetween(self, range)
    return self > range[1] and self < range[2]
end

function demoVec2Operators()
    local a = Vec2:new(3, 4)
    local b = Vec2:new(1, 2)
    local c = a + b
    local d = a - b
    local e = b * 3
    local neg = b:unaryMinus()
    local aGreater = a:compareTo(b) > 0
    local bSmaller = b:compareTo(a) < 0
    return c.x + d.x + e.x + neg.x + ((aGreater and 1 or 0)) + ((bSmaller and 1 or 0))
end

function demoInvokeOperator()
    local triple = Multiplier:new(3)
    return triple:invoke(7)
end

function demoExtensionFunction()
    local word = "hello"
    return string_shout(word)
end

function demoCustomInfix()
    local x = 5
    local inRange = int_isBetween(x, {0, 10})
    local notIn = int_isBetween(x, {5, 10})
    return ((inRange and 1 or 0)) + ((notIn and 1 or 0))
end

function demoWith()
    local v = Vec2:new(10, 20)
    return (function(self)
        return self.x + self.y
    end)(v)
end

function demoToMethod()
    local pair = {"key", 42}
    return {pair}
end

function demoCollectionOperators()
    local combined = ktox_plus({1, 2, 3}, {4, 5})
    local removed = ktox_minus({1, 2, 3, 4, 5}, 3)
    local m = ktox_mapPlus({["x"] = 1}, ({"y", 2}))
    local hasBoth = ktox_containsKey(m, "x") and ktox_containsKey(m, "y")
    return #combined + #removed + ((hasBoth and 1 or 0))
end

function demoMatches()
    local s = "hello123"
    local pattern = "\[a-z\]+\[0-9\]+"
    local dotForm = ktox_matches(s, pattern)
    local infixForm = ktox_matches(s, "\[a-z\]+\[0-9\]+")
    local noMatch = ktox_matches(s, "\[0-9\]+")
    return ((dotForm and 1 or 0)) + ((infixForm and 1 or 0)) + ((noMatch and 0 or 1))
end

function demoExtensionProperty()
    local phrase = "hello world foo"
    local wc = string_wordCount(phrase)
    local lc = string_lastChar(phrase)
    local single = "one"
    local wc2 = string_wordCount(single)
    local lc2 = string_lastChar(single)
    return wc + ((lc == "o" and 1 or 0)) + wc2 + ((lc2 == "e" and 1 or 0))
end

function demoExtensionFunctionOnUserDefined()
    local awesomeCat = CoolCat:new("teacup")
    awesomeCat:printName()
end

function demoClassExtensionProperty()
    local cat = CoolCat:new("Whiskers")
    return cat:greeting()
end

function demoCoolCatPlusAssign()
    local cat = CoolCat:new("Mittens")
    cat:plusAssign(10)
    cat:plusAssign(5)
    cat:minusAssign(3)
    return cat.coolPoints
end

function demoAllCompoundAssignOverloads()
    local acc = Accumulator:new(10)
    acc:plusAssign(5)
    acc:minusAssign(3)
    acc:timesAssign(2)
    acc:divAssign(4)
    acc:remAssign(5)
    return acc.value
end

